src/Snowflake/Command.php

```diff
@@ -149,18 +226,270 @@ public static function cloneDatabaseFromShared(Connection $connection, array $da
                         QueryBuilder::quoteIdentifier($schemaName),
                         QueryBuilder::quoteIdentifier($tableName),
                     ));
-                }
 
-                var_dump($schemaName);
-                if ($k > 9) {
-                    exit();
+                    foreach ($tableGrants as $tableGrant) {
+                        self::assignGrantToRole($connection, $tableGrant);
+                    }
                 }
             }
         }
     }
 
-    public static function exportUsersAndRoles(Connection $sourceSnflkConnection): array
+    public static function createRole(Connection $connection, array $role): void
+    {
+        if (isset($role['granted_by'])) {
+            self::useRole($connection, $role['granted_by']);
+        }
+
+        $connection->query(sprintf(
+            'CREATE ROLE %s',
+            $role['name']
+        ));
+
+        self::grantRoleToUser(
+            $connection,
+            (string) getenv('SNOWFLAKE_DESTINATION_ACCOUNT_USERNAME'),
+            $role['name']
+        );
+    }
+
+    public static function exportGrantsUsersAndRoles(Connection $connection, array $databases): array
+    {
+        $tmp = [];
+        foreach ($databases as $database) {
+            $roles = $connection->fetchAll(sprintf(
+                'SHOW ROLES LIKE %s',
+                QueryBuilder::quote($database . '%')
+            ));
```

Comment:
Tohle není správný předpoklad myslím. Ne všechny role mají takhle prefix... 

Z hlavy mě napadá https://github.com/keboola/connection/blob/master/legacy-app/application/src/Storage/Service/Backend/NameGenerator/SnowflakeObjectNameGenerator.php#L62

Tady by to mělo být právě tak, že si taháš ty jednotlivé resourcy (db, schemata, tabulky) a když najdeš v jejich grantech něco, co ještě v tom destination accountu nemáš, tak bys to měl vytvořit. 

Nebo špatně chápu, co to dělá

Reply:
vím... tady je spíš trochu nešťastně to `%` na 262 protože mě jde jen o hlavní role projektu a tím je jen `SAPI_1234`

role pak vytvářím podle ownershipu:
https://github.com/keboola/project-migration-tool/blob/ondra-init/src/Snowflake/Command.php#L123

co mě teď tak napadá že to není rekurzivní a pokud bude role třeba pod rolí `SAPI_1234_RO` nebo `SAPI_WORKSPACe_1234` tak ta se nevytvoří... ale jde o to jestli takový případ zanoření může nastat

Reply:
takže spíš jen tady odeberu to `%` ať to není zavádějící.... stejně se prochází jen ty hlavní role

Reply:
Jo, OK. Nebo prostě pojmenovat tu metodu tak, aby bylo jasné, že fakt tahá jen hlavní projektovou roli a je to expected.

Reply:
tak jsem to ještě prošel a upravil názvy metod aby víc odpovídali tomu co dělají... v principu jde o to stáhnout všechny granty a uživatele ze zdrojový databáze a to takhle:
1/ `SHOW ROLES LIKE SAPI_9472` -> odebral jsem to procento
2/ v metodě `getOtherRolesAndUsersToMainProjectRole()` donačtu všechny role a uživatele které vlastní hlavní role projektu - tj role worspacu, RO, share...
3/ donačtu všechny granty které patří jednotlivým rolím

tím bych měl mít všechny oprávnění a uživatele které jsou na daný projekt